Migrate to Azure Container Apps

mbarton opened the pull request:
We are seeing high p99 latency with the dGC API. The time appears to be lost between Azure API management and App Service. [Internal Discussion Link](https://forum.rcpch.tech/t/tpp-latency-using-downs-syndrome-charts/567/18).

Whilst we wait for Microsoft support this PR moves dGC to run on Container Apps, the same service that we run our audit platforms on. We don't know if this will improve latency but is at least a change we can make that will affect it somehow and continues to harmonise our deployments.

TODO

- [x] create staging instance
- [x] add deploy to staging to ci script and test
- [x] benchmark instance compared to app service instance
- [x] create live instance
- [x] add deploy to live to ci script and test
- [x] remove test branch from GitHub actions triggers and remove related federated credential
- [x] merge
- [x] change backend in API management to new live instance
- [x] update runbook [internal link](https://forum.rcpch.tech/t/digital-growth-charts-runbook/259)

mbarton committed:
Build and push container

mbarton committed:
Whoops forgot to login to the registry

mbarton commented on .github/workflows/all-branches-and-PRs-test.yml:
```diff
@@ -1,38 +0,0 @@
-    runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        python-version: ["3.9", "3.10", "3.11", "3.12"]
```
A build matrix makes sense for the underlying Python library but is not useful for a deployable like this - we only need test against the Python version it will be using in production

pacharanero replied:
Testing against newer Pythons can be useful for getting advance warning that a future Python version is going to break something in our code though... Agree that older/legacy Pythons have no additional value for a deployable like this. I suspect the Action was copied from a similar one used somewhere else. 

There might be the odd person self-hosting the API server that might find that text matrix useful, but this is not our target audience.

mbarton committed:
Remove old workflow file

mbarton committed:
Fix missing gunicorn

mbarton committed:
Try listening on all hosts

Getting a very opaque ACA error: upstream connect error or disconnect/reset before headers. retried and the latest reset reason: remote connection failure, transport failure reason: delayed connect error: 111.

I suspect it's because gunicorn is listening on the containers 127.0.0.1 so ACA ingress can't talk to it.

mbarton committed:
Deploy to staging

mbarton committed:
Oops use the right commands

mbarton committed:
Quick stress test script

mbarton commented:
I did some quick and dirty 10,000 request with no concurrency benchmarking (see `s/stress-test`). I wanted to try and make sure performance on the Azure Container Apps consumption tier wasn't noticeably worse than our existing app service instances. They come out about the same and well within the envelope for the API currently (never really more than hundreds of requests per minute).

## APP SERVICE

33.5774 requests per second

Slowest: 578.1 ms
Fastest: 19.2 ms

<img width="700" height="295" alt="app_service_histogram" src="https://github.com/user-attachments/assets/ec44259a-4674-4c89-aab9-f2ebdbc5632a" />


| Percentile | Latency |
|------------|---------|
| 10         | 25      |
| 25         | 27.4    |
| 50         | 28.8    |
| 75         | 30.3    |
| 90         | 33.8    |
| 95         | 37.8    |
| 99         | 63.7    |

## AZURE CONTAINER APPS

32.8361 requests per second

Slowest: 290.0 ms
Fastest: 20.2 ms
<img width="703" height="290" alt="Screenshot 2025-08-27 154748" src="https://github.com/user-attachments/assets/df698f57-2f1e-489b-9fbf-29721efe3151" />




| Percentile | Latency |
|------------|---------|
| 10         | 24.8    |
| 25         | 27.1    |
| 50         | 29.1    |
| 75         | 31.6    |
| 90         | 35.7    |
| 95         | 40.0    |
| 99         | 63.5    |

pacharanero commented:
> Fastest: 192.0 ms

Is this a typo? I think 19ms from the histogram

mbarton committed:
Add deploy to live

mbarton committed:
oops

mbarton committed:
Remove test branch

mbarton commented:
Seen on [LIVE](https://dgc-live.icywave-fab27df8.uksouth.azurecontainerapps.io) (merged by @mbarton 295 hours, 42 minutes and 7 seconds ago) Please check your changes!
